examples/07_tutorials/ai-reco-of-recos/utils/openai_utils.py

```diff
@@ -0,0 +1,456 @@
+    return model_info.client.embeddings.create(input=[text], model=model_info.model_name).data[0].embedding
+
+
+
+def call_llm(prompt: str, model_info: Union[MulitmodalProcessingModelInfo, TextProcessingModelnfo], temperature = 0.2, imgs=[]):
+    content = [{"type": "text", "text": prompt}]
+    content = content + prepare_image_messages(imgs)
+    messages = [
+        {"role": "user", "content": "You are a helpful assistant that processes text and images."},
+        {"role": "user", "content": content},
+    ]
+    
+    if model_info.client is None: model_info = instantiate_model(model_info)
+    # print(">>>>>>>>>>>>>>>>> call_llm model_info", model_info)
+
+    if (model_info.model_name == "gpt-4o") or ((model_info.model_name == "gpt-45")):
+        return call_4(messages, model_info.client, model_info.model, temperature)
+    elif model_info.model_name == "gpt-4.1":
+        return call_41(messages, model_info.client, model_info.model, temperature)
+    elif model_info.model_name == "o1":
+        return call_o1(messages, model_info.client, model_info.model, model_info.reasoning_efforts)
+    elif model_info.model_name == "o1-mini":
+        return call_o1_mini(messages, model_info.client, model_info.model)
+    elif model_info.model_name == "o3":
+        return call_o3(messages, model_info.client, model_info.model, model_info.reasoning_efforts)
+    elif model_info.model_name == "o3-mini":
+        return call_o3_mini(messages, model_info.client, model_info.model, model_info.reasoning_efforts)
+    elif model_info.model_name == "o4-mini":
+        return call_o4_mini(messages, model_info.client, model_info.model, model_info.reasoning_efforts)
+    else:
+        return call_4(messages, model_info.client, model_info.model, temperature)
+
+
+@retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(10))
+def call_4(messages, client, model, temperature = 0.2):
+    # print(f"\nCalling OpenAI APIs with {len(messages)} messages - Model: {model} - Endpoint: {client._base_url}\n")
+    result = client.chat.completions.create(model = model, temperature = temperature, messages = messages)
+    return result.choices[0].message.content
+
+@retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(10))
+def call_41(messages, client, model, temperature = 0.2):
+    print(f"\nCalling OpenAI APIs with {len(messages)} messages - Model: {model} - Endpoint: {client._base_url}\n")
+    result = client.chat.completions.create(model = model, temperature = temperature, messages = messages)
+    return result.choices[0].message.content
+      
+@retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(10))
+def call_o1(messages,  client, model, reasoning_effort ="medium"): 
+    # print(f"\nCalling OpenAI APIs with {len(messages)} messages - Model: {model} - Endpoint: {client._base_url}\n")
+    response = client.chat.completions.create(model=model, messages=messages, reasoning_effort=reasoning_effort)
```

Comment:
Please consider using Responses API if possible, for both reasoning and non-reasoning models.